packages/@uppy/react/src/DashboardModal.ts

```diff
@@ -0,0 +1,275 @@
+import {
+  cssSize,
+  locale,
+  metaFields,
+  plugins,
+  uppy as uppyPropType,
+} from './propTypes.ts'
+import getHTMLProps from './getHTMLProps.ts'
+import nonHtmlPropsHaveChanged from './nonHtmlPropsHaveChanged.ts'
+import type { DashboardProps } from './Dashboard.ts'
+
+interface DashboardModalProps<M extends Meta, B extends Body>
+  extends DashboardProps<M, B> {
+  animateOpenClose?: boolean
+  disableLocalFiles?: boolean
+  autoOpenFileEditor?: boolean
+  browserBackButtonClose?: boolean
+  showRemoveButtonAfterComplete?: boolean
+  showSelectedFiles?: boolean
+  theme?: 'auto' | 'dark' | 'light'
+  waitForThumbnailsBeforeUpload?: boolean
+  fileManagerSelectionType?: 'files' | 'folders' | 'both'
+  hideRetryButton?: boolean
+  showLinkToFileUploadResult?: boolean
+  closeAfterFinish?: boolean
+  disabled?: boolean
+  target?: PluginTarget<M, B>
+  open?: boolean
+  onRequestClose: () => void
+  closeModalOnClickOutside?: boolean
+  disablePageScrollWhenModalOpen?: boolean
+  hideCancelButton?: boolean
+  hidePauseResumeButton?: boolean
+}
+
+/**
+ * React Component that renders a Dashboard for an Uppy instance in a Modal
+ * dialog. Visibility of the Modal is toggled using the `open` prop.
+ */
+
+class DashboardModal<M extends Meta, B extends Body> extends Component<
+  DashboardModalProps<M, B>
+> {
+  static propTypes = {
+    uppy: uppyPropType.isRequired,
+    target:
+      typeof window !== 'undefined' ?
+        PropTypes.instanceOf(window.HTMLElement)
+      : PropTypes.any,
+    open: PropTypes.bool,
+    onRequestClose: PropTypes.func,
+    closeModalOnClickOutside: PropTypes.bool,
+    disablePageScrollWhenModalOpen: PropTypes.bool,
+    inline: PropTypes.bool,
+    plugins,
+    width: cssSize,
+    height: cssSize,
+    showProgressDetails: PropTypes.bool,
+    note: PropTypes.string,
+    metaFields,
+    proudlyDisplayPoweredByUppy: PropTypes.bool,
+    autoOpenFileEditor: PropTypes.bool,
+    animateOpenClose: PropTypes.bool,
+    browserBackButtonClose: PropTypes.bool,
+    closeAfterFinish: PropTypes.bool,
+    disableStatusBar: PropTypes.bool,
+    disableInformer: PropTypes.bool,
+    disableThumbnailGenerator: PropTypes.bool,
+    disableLocalFiles: PropTypes.bool,
+    disabled: PropTypes.bool,
+    hideCancelButton: PropTypes.bool,
+    hidePauseResumeButton: PropTypes.bool,
+    hideProgressAfterFinish: PropTypes.bool,
+    hideRetryButton: PropTypes.bool,
+    hideUploadButton: PropTypes.bool,
+    showLinkToFileUploadResult: PropTypes.bool,
+    showRemoveButtonAfterComplete: PropTypes.bool,
+    showSelectedFiles: PropTypes.bool,
+    waitForThumbnailsBeforeUpload: PropTypes.bool,
+    fileManagerSelectionType: PropTypes.string,
+    theme: PropTypes.string,
+    // pass-through to ThumbnailGenerator
+    thumbnailType: PropTypes.string,
+    thumbnailWidth: PropTypes.number,
+    locale,
+  }
+
+  // Must be kept in sync with @uppy/dashboard/src/Dashboard.jsx.
+  static defaultProps = {
```

Comment:
Shouldn't this `satisfies` some type?

Reply:
We have [`react/default-props-match-prop-types`](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/default-props-match-prop-types.md) ESLint rule that validates it already, I'm not sure we need more.

Reply:
AFAIK we are moving away from `defaultProps` and `propTypes` in favor of TS with default options. Wouldn't this be a good moment to do that?

Reply:
I don't know what this is, but I guess it would be a breaking change to do that on `main`, and that should happen in `4.x`, no?

Reply:
It's not a breaking change, prop types are only used in development before types were a thing. `defaultProps` is basically like `defaultOptions`. We no longer need prop types as types replace them. `defaultProps` could be replaced but it's more a matter of taste, where's prop types is completely redundant.

So at least `propTypes` should be removed in this PR as it's the TS PR.

Reply:
I don’t think that’s true, the README says it’s about runtime checks: https://www.npmjs.com/package/prop-types
Types are a dev time only type of check, not quite the same thing

Reply:
They are runtime checks for development, they generally are stripped from prod. Here is the recommendation from the React core team:

> We recommend using [TypeScript](https://www.typescriptlang.org/) instead of checking prop types at runtime — [react](https://react.dev/reference/react/Component#static-proptypes).

Let's remove them, it's double the work to maintain but less accurate then TS.

Reply:
It's not a breaking change in any way so it's fine either way 👍

Reply:
Yeah but folks who depends on that plugin won't get the types, so it seems better not to remove it just yet.